in-progress/8754-slashing-and staking.md

```diff
@@ -0,0 +1,357 @@
+
+Validators vote by signing the slashing proposal and gossiping back the signed message over the p2p. Anyone can submit a slashing proposal that has gathered enough signatures to the L1 for verification. 
+
+Based on the offense in question, the rollup contract establishes max slashable stakes. The rollup contract also verifies certain conditions have occured before executing the slash. 
+
+This is more subjective than purely onchain slashing but it enables precision that is not possible with purely onchain slashing. We can adopt Ethereum's principle of "honest validators should not be slashed" at the expense of increased coordination cost.
+
+
+### Users of PoG Slashing
+
+Validators are the users of the PoG slashing mechanism. Full nodes can contribute data to validators but they should not be able to vote for a slashing proposal. 
+
+
+## PoG Slashable Offences
+
+These are actions that can and will be slashed using PoG.
+
+### 1. Committee signs off on an invalid state root
+
+Some validators implement a `ALWAYS_EXECUTE` flag which causes them to execute every block regardless of whether they're in the committee or not. Validators who implement this flag are called "Executing Validators". 
+
+After every succesful block proposal, these Executing Validators will retrieve the list of `TxHash` from L1 and attempt to re-execute all transactions in the block. 
+
+In the happy path, the Executing Validators are able to obtain the `TxObjects` and proofs required to do so. If the resulting state root does not match what was published to L1, they read the list of all committee members who signed the invalid block. 
+
+The Executing Validator prepares a payload containing:
+1) The offense in question: Invalid state root
+2) The slot number of the invalid block.
+3) The list of committee members who signed off on the invalid block.
+
+and gossips through to the validator set. 
+
+Validators who receive this proposal first also read the L1 for the list of `TxHash` posted along the block in question. If it does not match what's in the slashing proposal, it discards the slashing proposal. 
+
+If the list of `TxHash` does match, they request all `TxEffects` and proofs from the p2p. As a last resort, they request from the validator who initiated the slashing proposal. They also read the state root posted to L1. 
+
+If they're able to retreive all the required data to re-xecute the block contents and find an invalid state root, they sign the slashing proposal and gossip the signed proposal via the p2p. 
+
+The executing validator who initialized the slashing proposal must aggregate signatures and post to L1 for verification once >50% of the stake has signed the proposal. 
+
+On the L1, the contract verifies that: 
+1) The epoch has indeed re-orged.
+2) The committee members named in the slashing proposal have signed the specific block in question.
+3) Validators in control of more than 50% of stake have signed the slashing proposal.
```

Comment:
If 50% of stake has signed off, do we need the other criteria?

As a technical aside, I'm not sure how we'd check that the block had the named committee members if we're also asserting that the block was reorged out.

Reply:
> As a technical aside, I'm not sure how we'd check that the block had the named committee members if we're also asserting that the block was reorged out.

I looked this up as I was unsure myself. This [post](https://ethresear.ch/t/clarification-on-how-calldata-persists-on-the-blockchain-and-how-optimistic-rollups-use-it/8136/9) says that calldata gets merkalized in contract storage. So someone needs to present to the contract, the calldata as well as a merkle proof to reference it. I couldn't find an implementation though.

Reply:
Im not entirely convinced the clarification in the link posted there is correct. The full calldata will be found in the transactions trie. As far as im aware there is no special location for calldata in a contracts storage, unless they are referring to a specific optimistic rollup implementation that stores the roots of the published calldata when performing batching.

for us to prove it's existence we will need to prove this against the block hash which is accesible in the evm. It is likely enough to just run off of signatures here rather than attempt to prove that the tx existed.